zigpy/device.py

```diff
@@ -47,118 +49,191 @@ class Device(zigpy.util.LocalLogMixin, zigpy.util.ListenableMixin):
+    def schedule_group_membership_scan(self) -> asyncio.Task:
         """Rescan device group's membership."""
-        if self._group_scan_handle and not self._group_scan_handle.done():
+        if self._group_scan_task and not self._group_scan_task.done():
             self.debug("Cancelling old group rescan")
-            self._group_scan_handle.cancel()
-        self._group_scan_handle = asyncio.ensure_future(self.group_membership_scan())
+            self._group_scan_task.cancel()
+
+        self._group_scan_task = asyncio.create_task(self.group_membership_scan())
+        return self._group_scan_task
 
     async def group_membership_scan(self) -> None:
         """Sync up group membership."""
         for ep_id, ep in self.endpoints.items():
             if ep_id:
                 await ep.group_membership_scan()
 
-    async def get_node_descriptor(self):
+    @property
+    def initializing(self) -> bool:
+        """Return True if device is being initialized."""
+        return self._initialize_task is not None and not self._initialize_task.done()
+
+    def cancel_initialization(self) -> None:
+        """Cancel initialization call."""
+        if self.initializing:
+            self.debug("Canceling old initialize call")
+            self._initialize_task.cancel()
+
+    def schedule_initialize(self) -> Optional[asyncio.Task]:
+        # Already-initialized devices don't need to be re-initialized
+        if self.is_initialized:
+            self.debug("Skipping initialization, device is fully initialized")
+            self._application.device_initialized(self)
+            return None
+
+        self.debug("Scheduling initialization")
+
+        self.cancel_initialization()
+        self._initialize_task = asyncio.create_task(self.initialize())
+
+        return self._initialize_task
+
+    async def get_node_descriptor(self) -> zdo.types.NodeDescriptor:
         self.info("Requesting 'Node Descriptor'")
-        try:
-            status, _, node_desc = await self.zdo.Node_Desc_req(
-                self.nwk, tries=2, delay=0.1
+
+        status, _, node_desc = await self.zdo.Node_Desc_req(
+            self.nwk, tries=2, delay=0.1
+        )
+
+        if status != zdo.types.Status.SUCCESS:
+            raise zigpy.exceptions.InvalidResponse(
+                f"Requesting Node Descriptor failed: {status}"
             )
-            if status == zdo.types.Status.SUCCESS:
-                self.node_desc = node_desc
-                self.info("Node Descriptor: %s", node_desc)
-                return node_desc
-            else:
-                self.warning("Requesting Node Descriptor failed: %s", status)
-        except (asyncio.TimeoutError, zigpy.exceptions.ZigbeeException):
-            self.warning("Requesting Node Descriptor failed", exc_info=True)
+
+        self.node_desc = node_desc
+        self.info("Got Node Descriptor: %s", node_desc)
+
+        return node_desc
+
+    async def initialize(self, *, tries=3, delay=0.5):
```

Comment:
Any reason retryable decorator doesn't work here anymore?

Reply:
Gotta emit `device_init_failure` (which is apparently unused) and call `self.application.remove` when it fails. I guess you could also attach a callback to the `Task` object created during the `schedule_initialize` method.

Do you think it would be better to keep around devices that failed to initialize so that a re-join has a higher chance of succeeding due to some info already being filled out?

Reply:
https://github.com/zigpy/zigpy/pull/746 it is an async callback for the retryable decorator

> Do you think it would be better to keep around devices that failed to initialize so that a re-join has a higher chance of succeeding due to some info already being filled out

I was thinking about that too. I'm onboard, just need to modify ZHA so it doesn't try to discover entities for not fully initialized devices

Reply:
`Application.remove` utilizes retryable requests with some pretty long timeouts. Should all requests to a device get cancelled if it leaves or re-joins? Or just the removal ones? While somewhat unlikely, it seems possible for a device to fail initialization, the removal taskgets started, and the device re-joins during a removal retry, causing it to be kicked off the network.

Alternatively, we could also *not* send any ZDO remove requests to the device when an initialization failure occurs, since a device that fails to initialize is likely already factory reset again and if the device miraculously starts sending requests again, continuing to initialize it seems like desired behavior?

Reply:
Good question. Sounds like `Application.remove` should cancel the `Device.initialize()` and `Device.initialize()` should cancel `device.remove()`. Although the timeouts are mostly waiting for a reply, meaning the request was sent out anyway. But we definitely should cancel any remaining retries.